<a href="https://colab.research.google.com/github/gulabpatel/AIAg/blob/main/GIS/Geospatial/21_Visualizing_LiDAR_Data_with_Leafmap.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

[![image](https://studiolab.sagemaker.aws/studiolab.svg)](https://studiolab.sagemaker.aws/import/github/opengeos/geoai/blob/main/docs/examples/dataviz/lidar_viz.ipynb)
[![image](https://img.shields.io/badge/Open-Planetary%20Computer-black?style=flat&logo=microsoft)](https://pccompute.westeurope.cloudapp.azure.com/compute/hub/user-redirect/git-pull?repo=https://github.com/opengeos/geoai&urlpath=lab/tree/geoai/docs/examples/dataviz/lidar_viz.ipynb&branch=main)
[![image](https://colab.research.google.com/assets/colab-badge.svg)](https://githubtocolab.com/opengeos/geoai/blob/main/docs/examples/dataviz/lidar_viz.ipynb)

# Visualizing LiDAR Data with Leafmap

This notebook demonstrates how to visualize LiDAR data using [leafmap](https://leafmap.org).

## Installation

Uncomment and run the following cell to install the required Python packages.

In [1]:
%pip install "leafmap[lidar]" open3d

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 10.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 420.5/420.5 MB 2.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 108.6/108.6 kB 13.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.9/2.9 MB 28.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 79.5 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 346.3/346.3 kB 31.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 66.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.4/10.4 MB 63.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 77.1/77.1 kB 9.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 139.4/139.4 kB 13.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.3/2.3 MB 

## Import libraries

In [2]:
import leafmap

## Download data

Download a [sample LiDAR dataset](https://drive.google.com/file/d/1H_X1190vL63BoFYa_cVBDxtIa8rG-Usb/view?usp=sharing) from Google Drive. The zip file is 52.1 MB and the uncompressed LAS file is 109 MB.

In [3]:
url = 'https://open.gishub.org/data/lidar/madison.zip'
filename = 'madison.las'

In [4]:
leafmap.download_file(url, 'madison.zip', unzip=True)

Downloading...
From: https://open.gishub.org/data/lidar/madison.zip
To: /content/madison.zip
100%|██████████| 54.7M/54.7M [00:00<00:00, 80.3MB/s]


Extracting files...


'/content/madison.zip'

## Metadata

Read the LiDAR data

In [5]:
las = leafmap.read_lidar(filename)

The LAS header.

In [6]:
las.header

<LasHeader(1.3, <PointFormat(1, 0 bytes of extra dims)>)>

The number of points.

In [7]:
las.header.point_count

4068294

The list of features.

In [8]:
list(las.point_format.dimension_names)

['X',
 'Y',
 'Z',
 'intensity',
 'return_number',
 'number_of_returns',
 'scan_direction_flag',
 'edge_of_flight_line',
 'classification',
 'synthetic',
 'key_point',
 'withheld',
 'scan_angle_rank',
 'user_data',
 'point_source_id',
 'gps_time']

## Read data

Inspect data.

In [9]:
las.X

array([5324343, 5324296, 5323993, ..., 5784049, 5784359, 5784667],
      dtype=int32)

In [10]:
las.Y

array([8035264, 8035347, 8035296, ..., 7550110, 7550066, 7550026],
      dtype=int32)

In [11]:
las.Z

array([36696, 34835, 34826, ..., 36839, 36858, 36842], dtype=int32)

In [12]:
las.intensity

array([ 9, 41, 24, ..., 87, 80, 95], dtype=uint16)

## PyVista

Visualize LiDAR data using the [pyvista](https://github.com/pyvista/pyvista) backend.

In [13]:
leafmap.view_lidar(filename, cmap='terrain', backend='pyvista')

This function is not supported in Google Colab.


![](https://i.imgur.com/xezcgMP.gif)

## ipygany

Visualize LiDAR data using the [ipygany](https://github.com/QuantStack/ipygany) backend.

In [14]:
leafmap.view_lidar(filename, backend='ipygany', background='white')

This function is not supported in Google Colab.


![](https://i.imgur.com/MyMWW4I.gif)

## Panel

Visualize LiDAR data using the [panel](https://github.com/holoviz/panel) backend.

In [15]:
leafmap.view_lidar(filename, cmap='terrain', backend='panel', background='white')

This function is not supported in Google Colab.


![](https://i.imgur.com/XQGWbJk.gif)

## Open3D

Visualize LiDAR data using the [open3d](http://www.open3d.org) backend.

In [16]:
leafmap.view_lidar(filename, backend='open3d')

This function is not supported in Google Colab.


![](https://i.imgur.com/rL85fbl.gif)